Src/WindowsServer/WindowsServer.Shared/Implementation/AppServiceEnvironmentVariableMonitor.cs

```diff
@@ -0,0 +1,77 @@
+﻿namespace Microsoft.ApplicationInsights.WindowsServer.Implementation
+{
+    using System;
+    using System.Collections.Generic;
+    using System.ComponentModel;
+
+    internal delegate void MonitoredAppServiceEnvVarUpdated();
+
+    /// <summary>
+    /// Utility to monitor the value of environment variables which may change 
+    /// during the run of an application. Checks the environment variables 
+    /// at regular set intervals.
+    /// </summary>
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    internal sealed class AppServiceEnvironmentVariableMonitor : EnvironmentVariableMonitor
+    {
+        // event raised whenever any of the environment variables being watched get updated
+        public MonitoredAppServiceEnvVarUpdated MonitoredAppServiceEnvVarUpdatedEvent;
+
+        // Default interval between environment variable checks
+        internal static readonly TimeSpan DefaultMonitorInterval = TimeSpan.FromSeconds(60);
+
+        // Default list of environment variables tracked by this monitor.
+        internal static IReadOnlyCollection<string> PreloadedMonitoredEnvironmentVariables = new string[]
+        {
+            "WEBSITE_SITE_NAME",
+            "WEBSITE_SLOT_NAME",
+            "WEBSITE_HOME_STAMPNAME",
+            "WEBSITE_HOSTNAME",
+            "WEBSITE_OWNER_NAME"
+        };
+
+        // singleton pattern, this is the one instance of this class allowed
+        private static readonly AppServiceEnvironmentVariableMonitor SingletonInstance = new AppServiceEnvironmentVariableMonitor();
```

Comment:
Is it possible to avoid this static initialization? If the app is not running on WebApp, then the entire effort of this class ctr() and reading multiple env vars is wasted and leading to increasing application startup time. This maybe very minimal but .net core apps could be running in containers and restarted very often.
Perhaps move the entire logic to a lazy initializer?
(Lets track it as separate issue)

Reply:
Agreed. Let's track it in a separate issue. I've added it as #875.